### PR TITLE
Fixes #3460.  `xxxtoxxx` consistency

### LINKED ISSUE #3029: Refactor `MouseEvent` and `MouseEventEventArgs` to simplify
Similar in spirit, but somewhat less scope, than #2927 

- [x] `New/OnMouseEvent` now uses just `MouseEvent`.
- [x] `New/OnMouseClicked` now uses just `MouseEvent`.
- [x] Better document/rename `OfX`/`OfY`
- [x] Move mouse stuff out of `Responder` and into `View` so `Responder` can be nuked
- [x] Add a `docs/mouse.md`
- [ ] Beef-up/refactor mouse unit tests
- [ ] Build `MouseBindings` per #https://github.com/gui-cs/Terminal.Gui/issues/3778
- [ ] Merge MouseEvent and MouseEventEventArgs into a single class.
- [ ] Engineer something that removes the need for `Responder.IsOverridden` which breaks all rules of OO design.
- [ ] Ensure all library code uses consistent naming/usage
- [ ] Ensure all Scenario code uses best practices related to new design.
- [ ] Ensure all API documentation is awesome

### LINKED ISSUE #3460: Mixing signatures for ScreenToX between `int` and `Point`
`ScreenToContent` uses `Point`
`ScreenToViewport` uses `int`,`int`
`ScreenToFrame` uses `int`,`int`

image

Is it possible to please standardize these? or is there a reason they are not consistent. 

It makes debugging screen pointer issues difficult if you are trying to switch between the two methods to see which is more correct.

### PR DESCRIPTION
**Pull request by tig**
## Fixes

- Fixes #3460
- Partially Fixes #3029 (`X/Y` -> `Position`)

## Proposed Changes/Todos

- [x] `ScreenToFrame` -> takes `Point`
- [x] `ScreenToViewport` -> takes `Point`

### CODE REVIEW

**Comment by dodexahedron** [comment]
Hey quick easy optiization:

Make them take it as `in`, too, so reference can be passed by default without needing to be explicitly done at the callsite.

**Comment by dodexahedron** [comment]
That, by the way, is just a good habit to get into when passing structs - especially blittable ones.

**Comment by dodexahedron** [comment]
For certain other cases, there's also now `ref readonly` for method parameters, in c# 12, but that is _slightly_ different. MS has a matrix somewhere showing the effects and compatibility at the callsite for ref, ref readonly, in, and out, for the curious. I think it's linked from the what's new for c# 12.

**Comment by tig** [comment]
`in` -> done.

**Comment by dodexahedron** [comment]
Hm. Why builing new points in a lot of those lines rather than just passing the point itself?

File: Terminal.Gui/Application.cs [hunk 1]
```diff
@@ -1556,7 +1556,7 @@ internal static void OnMouseEvent (MouseEvent mouseEvent)
             return;
         }
 
-        var view = View.FindDeepestView (Current, mouseEvent.X, mouseEvent.Y);
+        var view = View.FindDeepestView (Current, new (mouseEvent.X, mouseEvent.Y));
```
Comment (dodexahedron) [inline Terminal.Gui/Application.cs:1559]: Why new?

Better to make the MouseEventArgs itself have a point and pass that by ref.

Comment (dodexahedron) [reply Terminal.Gui/Application.cs:1559]: Can be done non-breaking by just adding the property and modifying the existing properties to update the underlying Point.

File: Terminal.Gui/View/Adornment/Border.cs [hunk 2]
```diff
@@ -273,7 +273,7 @@ protected internal override bool OnMouseEvent (MouseEvent mouseEvent)
 
             // Only start grabbing if the user clicks in the Thickness area
             // Adornment.Contains takes Parent SuperView=relative coords.
-            if (Contains (mouseEvent.X + Parent.Frame.X + Frame.X, mouseEvent.Y + Parent.Frame.Y + Frame.Y))
+            if (Contains (new (mouseEvent.X + Parent.Frame.X + Frame.X, mouseEvent.Y + Parent.Frame.Y + Frame.Y)))
```
Comment (dodexahedron) [inline Terminal.Gui/View/Adornment/Border.cs:276]: Hm. I'm pretty sure offsetting one of the points by the other two would be less implicit copying.

File: Terminal.Gui/View/Layout/ViewLayout.cs [hunk 3]
```diff
@@ -147,25 +146,22 @@ public virtual Rectangle FrameToScreen ()
...
-    /// <param name="x">Screen-relative column.</param>
-    /// <param name="y">Screen-relative row.</param>
-    public virtual Point ScreenToFrame (int x, int y)
+    /// <param name="location">Screen-relative coordinate.</param>
+    public virtual Point ScreenToFrame (in Point location)
     {
         if (SuperView is null)
         {
-            return new (x - Frame.X, y - Frame.Y);
+            return new (location.X - Frame.X, location.Y - Frame.Y);
```
Comment (dodexahedron) [inline Terminal.Gui/View/Layout/ViewLayout.cs:154]: Offset here is also a good idea.

**Review by dodexahedron — CHANGES_REQUESTED** [review]
Mostly same minor comments elsewhere, so I won't repeat it all over the place.

Otherwise, YAY!

**Comment by dodexahedron** [comment]
There's also an Offset overload that takes Point, too, which is better when you already have one.

File: Terminal.Gui/Application.cs (hunk 1 above)
Comment (dodexahedron) [reply Terminal.Gui/Application.cs:1559]: Actually, thinking about it a bit more...

If the second parameter to FindDeepestView were `ref`, you can either supply one if you already have one or create it in the FindDeepestView method.

`out` also can do that, but it's a bit non-obvious and not recommended to consume an `out` param, even though it's legal.

And `ref` has the advantage it won't _let_ you create an ephemeral value at the callsite.

Lots of other options, too, but key is just trying to minimize how many get created and thrown away.

**Comment by tig** [comment]
> Hm. Why building new points in a lot of those lines rather than just passing the point itself?

You mean in the tests? 

> At the first place in a method that assigns it to something, it'll still get value copied anyway.
> 
> And with a non-referenceable ephemeral new() in a method call, it will be copied twice, making the `in` not apply.

I was focused on getting the API right; perf next.

image

How do you suggest I address this?

File: Terminal.Gui/View/Layout/ViewLayout.cs (hunk 3 above)
Comment (tig) [reply Terminal.Gui/View/Layout/ViewLayout.cs:154]: image

**Comment by tig** [comment]
I bit the bullet and changed `MouseEvent` to use `Point`.

**Comment by dodexahedron** [comment]
Sounds like it's probably pretty great.

I'm only half awake after a wrong number caller woke me up, so I'm useless for review right now but I'll get on that as soon as I'm up and running tomorrow. :)

**Pull request merged**